### PR TITLE
feat: CLIN-3090 save nextflow log file and add nf-test unit tests

### PR DESCRIPTION
**Pull request by LysianeBouchard**
## Description

In this PR, we introduce the following changes:

1. **Log File Copying**: Added code in the workflow.onComplete handler to copy the Nextflow log file to the output directory.
2. **Unit testing with nf-test**: Introduced nf-test for unit testing the log file copying logic
3. **Github workflow for nf-test**:  Added a github workflow to automatically run nf-test on pull requests and pushes to the main branch. To optimize performance, we only execute tests relevant to the modified files.
4. **nf-core subworkflows updated**: Updated the following subworkflows as recommended by the linter: 
  - `utils_nextflow_pipeline`
  - `utils_nfcore_pipeline`
5. **Documentation update**: Updated .github/CONTRIBUTING.md  

## Limitations

Here are a few limitations of this implementation regarding the saving of Nextflow logs:

- Logs generated after the `onComplete` handler will not be included in the copied log file. These logs are generally not essential for debugging.
- If an early error occurs, such as a parameter validation error, the `onComplete` handler is not triggered, and the log file is not copied. However, we think that the information displayed in stdout should be sufficient and detailed enough for troubleshooting in this case.

## Test description

Linter:

- Run `nf-core lint`: there are actually two warnings less because we updated nf-core subworkflows.

Unit tests:
- Run nf-test unit tests: ` nf-test test --verbose --tag local  --changed-since=main`
It run the new tests that I added as expected.

Testing log file writing corner cases locally: 
- Run the pipeline in stub mode, with and without log option and check that the nextflow log file is copied successfully.
`nextflow run main.nf -stub -profile test,docker`
`nextflow -log foo.log run main.nf -stub -profile test,docker`
- Run the pipeline in stub mode and press CTL-C to cancel the execution. Check that the log file is copied correctly.
- Remove the pipeline without the stub mode to trigger an error. Check that the log file is copied correctly.
- Run the pipeline in stub mode without specifying a valid profile to trigger an parameter validation error.
In this case, workflow handlers (both on complete and on error) are not executed and thus the log file is not copied.  The message available on stdout, however, should be enough to debug in this case. I don't think that it is necessary to add code to support this scenario.

- Test with real public data locally:
`nextflow -c data-test/test.config  run main.nf  -params-file data-test/params.json  -profile test,docker --input data-test/testSampleSheet.csv`

- Test on juno with s3 location for the output directory:
`nextflow -c  /root/nextflow/config/fusion.config   -c data-test/test.config  run Ferlab-Ste-Justine/Post-processing-Pipeline -r feat/CLIN-3090-save-nextflow-log-file  -params-file data-test/paramsJuno.json  -work-dir 's3://cqdg-prod-file-scratch/nextflow/scratch/lysiane'`
The log file is correctly copied.

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/ferlab/postprocessing/tree/master/.github/CONTRIBUTING.md)
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] Usage Documentation in `docs/usage.md` is updated.
- [ ] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [ ] `README.md` is updated (including new tool citations and authors/contributors).

### CODE REVIEW

File: .github/workflows/ci-nf-test.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+
+jobs:
+  test:
+    name: nf-test, NF ${{ matrix.NXF_VER }}
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        NXF_VER:
+          - "23.10.1"
```
Comment (LysianeBouchard) [inline .github/workflows/ci-nf-test.yml:18]: Since the tests tend to be slow, I only run them with the expected nextflow version in production. 
However, it is possible to add more nextflow versions if necessary.

File: .github/workflows/ci-nf-test.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+          wget -qO- https://code.askimed.com/install/nf-test | bash
+          sudo mv nf-test /usr/local/bin/
+
+      - name: Disk space cleanup
+        uses: jlumbroso/free-disk-space@54081f138730dfa15788a46383842cd2f914a1be # v1.3.1
+        
+      - name: Run nf-test
+        run: |
+          nf-test test \
+            --tag=local \
```
Comment (LysianeBouchard) [inline .github/workflows/ci-nf-test.yml:41]: I've tagged the tests that I wrote with "local" and added this filter here to ensure that `nf-test` only executes unit tests applicable to local resources. Let me know if you are ok with this convention. We can relax the tag filter later if we want to execute more tests.

File: .github/workflows/ci-nf-test.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            --changed-since="HEAD^1" \
```
Comment (LysianeBouchard) [inline .github/workflows/ci-nf-test.yml:42]: Here nf-test will only run tests applicable to files that changed since the specified revision. Here this should point on the last commit on the main branch before merging the pull request.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -18,8 +18,6 @@ concurrency:
 jobs:
   test:
     name: Run pipeline with test data
-    # Only run on push if this is the nf-core dev branch (merged PRs)
-    if: "${{ github.event_name != 'push' || (github.event_name == 'push' && github.repository == 'ferlab/postprocessing') }}"
```
Comment (LysianeBouchard) [inline .github/workflows/ci.yml:22]: I don't think that this apply to us.

File: subworkflows/local/utils_nfcore_postprocessing_pipeline/tests/main.function.nf.test
Comment (FelixAntoineLeSieur) [inline subworkflows/local/utils_nfcore_postprocessing_pipeline/tests/main.function.nf.test:1]: Faire un fichier spécifique pour cette fonction? Sinon ça va grossir rapidement quand on aura d'autres fonctions à tester dans ce même fichier

Comment (LysianeBouchard) [reply subworkflows/local/utils_nfcore_postprocessing_pipeline/tests/main.function.nf.test:1]: Oui bon point. C'est fait.

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -75,7 +74,8 @@ If you wish to contribute a new step, please use the following coding standards:
 6. Add sanity checks and validation for all relevant parameters.
 7. Perform local tests to validate that the new code works as expected.
 8. If applicable, add a new test command in `.github/workflow/ci.yml`.
-9. Add a description of the output files if relevant to `docs/output.md`.
+9. If applicable, write nf-test unit tests
+10. Add a description of the output files if relevant to `docs/output.md`.
```
Comment (FelixAntoineLeSieur) [inline .github/CONTRIBUTING.md:78]: Peut-être qu'on pourrait ajouter les logs qu'on produit au output.md éventuellement?

**Review by FelixAntoineLeSieur — DISMISSED** [review]

**Review by FelixAntoineLeSieur — APPROVED** [review]
Merci!
